addons/binding/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/handler/AccountHandler.java

```diff
@@ -0,0 +1,554 @@
+                if (!currentConnection.getIsLoggedIn()) {
+                    try {
+
+                        // read session data from property
+                        String sessionStore = this.stateStorage.findState("sessionStorage");
+
+                        // try use the session data
+                        if (!currentConnection.tryRestoreLogin(sessionStore)) {
+                            // session data not valid -> login
+                            int retry = 0;
+                            while (true) {
+                                try {
+                                    currentConnection.makeLogin();
+                                    break;
+                                } catch (ConnectionException e) {
+                                    // Up to 2 retries for login
+                                    retry++;
+                                    if (retry >= 2) {
+                                        currentConnection.logout();
+                                        throw e;
+                                    }
+                                    // give amazon some time
+                                    try {
+                                        Thread.sleep(2000);
+                                    } catch (InterruptedException exception) {
+                                        // throw the original exception
+                                        throw e;
+                                    }
+                                }
+                            }
+                            // store session data in property
+                            String serializedStorage = currentConnection.serializeLoginData();
+                            this.stateStorage.storeState("sessionStorage", serializedStorage);
+                        }
+                        this.connection = currentConnection;
+                    } catch (UnknownHostException e) {
+                        loginIsValid = false;
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Unknown host name '"
+                                + e.getMessage() + "'. Maybe your internet connection is offline");
+                    } catch (IOException e) {
+                        loginIsValid = false;
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                                e.getLocalizedMessage());
+                    } catch (URISyntaxException e) {
+                        loginIsValid = false;
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                                e.getLocalizedMessage());
+                    }
+                    if (loginIsValid) {
+                        handleValidLogin();
+                    }
+                }
+            }
+        } catch (Exception e) {
+            logger.error("check login fails {}", e);
```

Comment:
This can also happen later at runtime, if amazon requires to enter a captcha for the session renew. Should I change it anyway to debug?

Reply:
Yes, error should really only be used if there is something going unexpectedly wrong in the system, but not as a means of communication to the user.

Reply:
```
  } catch (HttpException | JsonSyntaxException e) {
            logger.debug("check login fails {}", e);
        } catch (Exception e) { // this handler can be removed later, if we know that nothing else can fail.
            logger.error("check login fails with unexpected error {}", e);
        }
```
This would be ok? I really don't know what can happen if the server returns an complete unexpected answer. In the past I had null point violations because an expected json element was not filled -> I have fixed that. At amazon.com (and only there) there was a redirect to http instead of https which causes an invalid cast exception -> I have fixed that also, but I do not know what else can happen.

Reply:
Yes, sure, that makes sense this way.